### ISSUE TITLE
Double ESC should be enabled for the every filter field

### ISSUE THREAD

**Issue by tierninho** [desktop/desktop#6369]
In the merge dialogue modal and clone modal dialogue, if the cursor is in the filter field **and** text is present, the `esc` should clear filter not close the modal. Hitting `esc` again should close the modal.

<img width="504" alt="screen shot 2018-12-07 at 3 53 25 pm" src="https://user-images.githubusercontent.com/14828183/49678298-4c4c9500-fa38-11e8-8539-39d6b9d93f0f.png">

Repro: 
- select Branch > Merge into current branch
- File > Clone Repo

Should be applied to Github, Enterprise tabs, but unsure of URL tab.
<img width="472" alt="screen shot 2018-12-10 at 11 28 12 am" src="https://user-images.githubusercontent.com/14828183/49756414-e433c400-fc6e-11e8-9a85-dc40fac962c7.png">

related to: https://github.com/desktop/desktop/pull/6236

cc @altaf933 in case you want to give it a go. 

1.5.1-beta2, mac

**Comment by altaf933** [comment]
Yeah, sure

**Comment by billygriffin** [comment]
Just wanted to capture feedback from @niik's review on #6531 here:

Hey @altaf933, thanks so much for wanting to contribute!

While the solution proposed in this PR solves the problem stated in the original issue it unfortunately comes with a few, unwanted, side-effects. The `onDismissed` callback is called for any user action that should lead to the dialog being dismissed, not just the <kbd>Escape</kbd> key being pressed but also clicking outside of the dialog, clicking the `X` icon at the top right corner of the dialog, or (on macOS) hitting <kbd>⌘</kbd><kbd>W</kbd>. That means that if there's text in the filter textbox and a user clicks the `X` the dialog will clear the filter text instead of closing as is the intended result.

I think we should refocus our attention on solving this in a way that is applicable to all dialogs that may contain a filter textbox. If we trace the chain of events from someone hitting <kbd>Escape</kbd> in a filter textbox (i.e. a `TextBox` component with its [`type` attribute](https://github.com/desktop/desktop/blob/f4efc0af4ef610f9abcc91403aa8f5547282fd0f/app/src/ui/lib/text-box.tsx#L45) set to `search`) we'll start in the [`onKeyDown` event](https://github.com/desktop/desktop/blob/f4efc0af4ef610f9abcc91403aa8f5547282fd0f/app/src/ui/lib/text-box.tsx#L233-L265) which has logic for determining whether the textbox should be cleared or not. If the textbox is not empty it will be cleared and the event will be marked as suppressed from performing it's default action (`e.preventDefault()`).

If we then move to the `<Dialog>` component and look at where the <kbd>Escape</kbd> key is handled there we'll again find it in the [`onKeyDown` method](https://github.com/desktop/desktop/blob/f4efc0af4ef610f9abcc91403aa8f5547282fd0f/app/src/ui/dialog/dialog.tsx#L317-L322).

I propose that in this method we check to see whether the keyboard event we're receiving has its `preventDefault` flag set or not. That way we should automatically enable the behavior we want for all filter textboxes inside of dialog component.

Now, unfortunately there's one complicating factor here. The Dialog component is a wrapper component around the native `<dialog>` html element. When we initially implemented it there was little to no support in React for it and as such we couldn't subscribe to the keydown event as we normally do in other components by adding a `onKeyDown={this.onKeyDown}` attribute and opted instead to manually subscribe to the event through [`addEventListener`](https://github.com/desktop/desktop/blob/f4efc0af4ef610f9abcc91403aa8f5547282fd0f/app/src/ui/dialog/dialog.tsx#L307). We'll need to change that to the conventional way now that the React TS definitions have somewhat caught up or else the timing between when these events are received will be off and cause us even more trouble.

So, in summary, I propose we do the following

1. Remove the manual subscription and unsubscription calls for the keydown event in the `Dialog` component and replace it with a conventional React subscription.
2. Inside of the onKeyDown event in the `Dialog` component we'll add a check to see whether the `defaultPrevented` flag of the event is set or not. If it is we'll return out of the method early.

I apologize for the wall of information, we'd be delighted if you would like to try your hand at implementing the solution I proposed above and please let us know if there's any part that's unclear and I'll try my best to clarify.

**Comment by tierninho** [comment]
This also affects the Rebase modal.

**Comment by shiftkey** [comment]
@tierninho can you confirm which steps of the flow you're referring to?

**Comment by tierninho** [comment]
Steps are for either the merge or rebase model are:

- open Branch menu, select either `Merge into Current Branch` or `Rebase current branch`
- type something in the filter
- hit `esc` key

Result: Entire modal closes.

Expected: Filter is cleared. If user hits `esc` key again, then entire modal should close.

1.7.0beta5, mac
